### PR TITLE
Solve the IK velocity step via JJ^T eigendecomposition, not JacobiSVD

### PR DESCRIPTION
**Pull request by hugogo1998**
### Description

`ChainIkSolverVelMimicSVD::CartToJnt()` runs once per Newton iteration, and profiling showed that most of its runtime was spent computing a thin `JacobiSVD` of the 6xN Jacobian.

This replaces that decomposition with the equivalent minimum-norm solve

```text
x = J^T (JJ^T)^# vin
```

using a self-adjoint eigendecomposition of the smaller `rows x rows` matrix `JJ^T`. The helper applies the same relative singular-value threshold as the existing SVD solve.

### Results

Measured on a 7-DOF Panda chain:

| workload | current | this change | speedup |
|---|---:|---:|---:|
| decomposition, 6x7 | 5.33–5.38 us | 1.85 us | ~2.9x |
| decomposition, 3x7 | 1.24–1.27 us | 0.76 us | ~1.6x |
| full IK solve, original perturbation | 1774.9 us | 1446.4 us | 1.23x |
| full IK solve, 6-DOF targets | 2173.2 us | 1654.0 us | 1.31x |

### Validation

Added tests comparing the helper against Eigen's thin `JacobiSVD` for 6x7 and 3x7 Jacobians. They cover full-rank, rank-deficient, ill-conditioned, and threshold-boundary cases.

- `clang-format` run on all changed C++ files
- 4 pseudoinverse tests pass in the MoveIt Humble CI image with `-Wall -Wextra -Werror`

### Checklist

- [x] Code is formatted using `clang-format`
- [x] No tutorial or documentation changes are needed
- [x] No user-facing API change; no `MIGRATION.md` entry is needed
- [x] Added tests that compare the new solve with the existing thin-SVD result
- [x] No GUI changes
- [ ] Review another open pull request while waiting for review

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved inverse kinematics solving for weighted and mimic-joint configurations, including rank-deficient and ill-conditioned systems.
  * Added threshold-based handling for near-zero singular values to improve numerical stability.

* **Tests**
  * Added automated coverage for full-rank, rank-deficient, threshold-boundary, and ill-conditioned pseudoinverse calculations.

### CODE REVIEW

**Review by rhaschke — COMMENTED** [review]
Did you try computing a [**thin** JacobiSVD](https://libeigen.gitlab.io/eigen/docs-5.0/classEigen_1_1JacobiSVD.html) alternatively?
`JacobiSVD<MatrixXf, ComputeThinU | ComputeThinV>`

Manually computing the pseudoinverse seems to be awkward.

**Comment by hugogo1998** [comment]
The existing solver is already configured with `ComputeThinU | ComputeThinV`, so the baseline in the PR is the thin JacobiSVD implementation.

You’re right that having the pseudoinverse calculation inline made `CartToJnt()` harder to follow. I moved it into a small private helper and added tests comparing it with the existing thin-SVD solve. The tests cover full-rank 6x7 and 3x7 matrices, rank-deficient and ill-conditioned matrices, and singular values immediately above and below the configured cutoff.

I also reran the benchmark. For 6x7 matrices, thin JacobiSVD took 5.33–5.38 µs and the eigendecomposition about 1.85 µs. For 3x7, the numbers were 1.24–1.27 µs and about 0.76 µs. Both implementations selected the same effective rank in all tested cases.

**Comment by rhaschke** [comment]
> The existing solver is already configured with `ComputeThinU | ComputeThinV`

You are right. 

> You’re right that having the pseudoinverse calculation inline made `CartToJnt()` harder to follow. I moved it into a small private helper and added tests comparing it with the existing thin-SVD solve.

I cannot find those recent changes. Did you push them already?

**Comment by hugogo1998** [comment]
Sorry for overlooking that. They are on the PR now in f7cd23fdbd7135d98f9f5125e3ea33848c565f0d. I moved the pseudoinverse calculation into a small internal helper and added comparisons against Eigen’s thin JacobiSVD for 6x7 and 3x7 matrices, including full-rank, rank-deficient, ill-conditioned, and threshold-boundary cases.

**Comment by hugogo1998** [comment]
Hi @rhaschke, could you pls take a look again when you get time? Please let me know if I overlook something, and any feedback is appreciated, thanks

**Comment by hugogo1998** [comment]
Hi @rhaschke I have pushed the changes as we discussed. Could you take another look when you get time? I'd appreciate it if you have any feedback, thanks

**Review by rhaschke — APPROVED** [review]

**Comment by hugogo1998** [comment]
Hi @rhaschke could you queue this pull request so that I can close it? Thanks

**Comment by rhaschke** [comment]
> Hi @rhaschke could you queue this pull request so that I can close it? Thanks

I was actually hoping for a second review by an actual MoveIt2 maintainer.

**Comment by rhaschke** [comment]
I suggest removing the `svd_` instance as well. Could you please provide a commit message text motivating this change? I even suggest to include this motivation in the pseudoinverse.hpp file.

**Comment by hugogo1998** [comment]
Thanks, addressed in c7eddb26b. I removed the `svd_` instance and now store the position-only mode and threshold directly, while taking the active-joint count from the reduced Jacobian. I also updated the class description and documented in `pseudoinverse.hpp` why the row-space eigendecomposition is used. The commit message includes the same motivation. The pseudoinverse tests pass, and the changed solver source compiles with `-Wall -Wextra -Werror`. Could you please take another look?

File: moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp [hunk 1]
```diff
@@ -111,19 +111,20 @@ int ChainIkSolverVelMimicSVD::CartToJnt(const JntArray& q_in, const Twist& v_in,
   vin.topRows<3>() = Eigen::Map<const Eigen::Array3d>(v_in.vel.data, 3) * cartesian_weights.topRows<3>().array();
   vin.bottomRows<3>() = Eigen::Map<const Eigen::Array3d>(v_in.rot.data, 3) * cartesian_weights.bottomRows<3>().array();
 
-  // Do a singular value decomposition: J = U*S*V^t
-  svd_.compute(jac.topRows(rows));
+  const auto jacobian = jac.topRows(rows);
+  const Eigen::VectorXd x_sol =
+      kdl_kinematics_plugin::internal::solvePseudoinverse(jacobian, vin.topRows(rows), threshold_);
```
[the remark this answers is not in the document]
Comment (hugogo1998) [reply moveit_kinematics/kdl_kinematics_plugin/src/chainiksolver_vel_mimic_svd.cpp]: Thanks, this is a valid limitation of the normal-equation approach. However, the solver defaults to a threshold of 1e-3, and all in-tree call sites use that default. At this threshold, modes affected by the JJᵀ precision limit are already discarded.
Adding an SVD fallback only for thresholds below approximately 1e-8 would add complexity to an edge case not exercised in-tree. I would prefer to keep the current implementation unless maintainers require behavioral parity for arbitrary user-supplied thresholds.
